### DISCUSSION TITLE
A Question about injecting languages for syntax highlighting

### DISCUSSION THREAD

**Discussion by FoamScience** [tree-sitter/tree-sitter#1493] [Q&A]
Hi again,

For this piece of code:
```
code
#{
    /* Comment #{#} */
    /* Another comment */
    pointField points(30);
    points[0]  = point(0.5, 0, -0.5);
#};
```
My parser generates the following tree; Note that the first comment has "#{" inside it, and it gets parsed as a comment in my grammar.
The second one (the one with no "#{" or "#}") is considered external code (to the grammar) and is parsed as part of `code_body`
```
(source_file [0, 0] - [7, 0]
  (key_value [0, 0] - [6, 3]
    keyword: (identifier [0, 0] - [0, 4])
    value: (code [1, 0] - [6, 2]
      (comment [2, 0] - [2, 22])
      (code_body [3, 0] - [5, 37]))))
```

Now I want to highlight the content of `code_body`, so I put the following in `queries/injections.scm`:
```
; Parse dynamic code as with c++ parser
(code (code_body) @injection.content (#set! injection.language "cpp"))
```

But that does not seem to work on the command line; Does the CLI client support language injections? or am I doing something wrong?

PS: `ts dump-languages` reports my grammar in addition to "c" and "cpp" in case these are needed for the cli to work

**Comment by ahelwer** [comment]
I actually don't know whether the CLI supports language injection, but [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) certainly does; I found that plugin quite useful for prototyping my tree-sitter grammar, especially since nobody actually uses `tree-sitter highlight` for anything other than testing as far as I know. You can set up nvim-treesitter as a plugin with symlinks to your grammar repo for local prototyping, you don't need to go through their release process.

**Reply by FoamScience** [reply]
Thanks! I think that's way better than viewing with the CLI too :)

**Reply by ahelwer** [reply]
The only thing to watch out for is that nvim-treesitter doesn't give great error messages for malformed queries. So if you get an error switch back to using the CLI.

**Reply by alemuller** [reply]
You can use `:go <loc>` in the query to jump to the error using the position provided by nvim's error msg.

Be aware that nvim and `tree-sitter-cli` applies the captures in different orders. IIRC nvim applies all captures in the order they appear of the query, while `tree-sitter-cli` only applies the first (or the last).

By the way, nvim has a cool feature that lets you take a cli screenshot of the current window.
On nvim:
`:call nvim__screenshot("filename")`
Then, on terminal:
`$ cat filename`

Do save your files before calling `nvim__screenshot`, it segfaults every once and a while.

**Reply by FoamScience** [reply]
Hmm 
![2021-11-20_17-03](https://user-images.githubusercontent.com/34474472/142733013-ef30c215-6660-44bd-ab09-c98b077c685e.png)

Even with nvim-treesitter, there is no highlighting of the code block!
This is how it looks with `ts highlight`:
![2021-11-20_17-04](https://user-images.githubusercontent.com/34474472/142733061-6b75cb90-22c6-453f-ae3a-72c45d8301d2.png)

Any ideas?

**Reply by alemuller** [reply]
Have you tried `@cpp` capture on nvim?

**Reply by FoamScience** [reply]
OH hell, that magically worked; I wasn't aware of such a capture. Thanks!

**Reply by FoamScience** [reply]
Hey guys,

My parser parses relatively complex expressions as a single node:
```
div(1|p,U)  on;
```
parses as:
```
(identifier) (boolean)
```

Now for syntax-highlighting the identifier; I'm cheating with injections:
```
((identifier) @lua)
```

But that will sometimes generate errors, or at least do unnecessary highlighting (eg. all caps variable names will be highlighted by lua parser but I don't want that)

Is there a quick hack to highlight just "punctuation" inside these identifier nodes?
